File: Minimax/375. Guess Number Higher or Lower II/Solution.java
class Solution {
    public int getMoneyAmount(int n) {
        int[][] dp=new int[n+1][n+1];
        for (int j=1; j<=n; j++){
            // decrement because cost=x+max(cost[i:x-1], cost[x+1:j])
            // we will need to know cost[x+1:j]
            for (int i=j-1; i>0; i--){
                if (i+1==j){
                    //when only two numbers
                    dp[i][j]=i;
                }else{
                    //at least 3 numbers
                    int cost=Integer.MAX_VALUE;
                    for (int x=i+1; x<j; x++){
                        int cur=x + Math.max(dp[i][x-1], dp[x+1][j]);
                        cost=Math.min(cost, cur);
                    }
                    dp[i][j]=cost;
                }
            }
        }
        return dp[1][n];
    }
}

File: Minimax/375. Guess Number Higher or Lower II/Solution.py
def getMoneyAmount(self, n: int) -> int:
        dp=[[0 for j in range(n+1)] for i in range(n+1)]
        def mini_max(i, j):
            if j<=i:
                return 0
            if dp[i][j]:
                return dp[i][j]
            cost=float('inf')
            for x in range(i, j+1):
                cur = x+max(mini_max(i, x-1), mini_max(x+1, j))
                cost=min(cost, cur)
            dp[i][j]=cost
            return cost
        return mini_max(1, n)